services/sonar/sonar-generic.service.js

```diff
@@ -0,0 +1,152 @@
+'use strict'
+
+const { metric } = require('../text-formatters')
+const SonarBase = require('./sonar-base')
+const { patternBase, queryParamSchema, getLabel } = require('./sonar-helpers')
+
+// This service is intended to be a temporary solution to avoid breaking
+// any existing users/badges that were utilizing the "other" Sonar metrics
+// with the Legacy Shields service implementation for Sonar badges.
+// The legacy implementation simply rendered a brightgreen badge with the value,
+// regardless of what the value actually was.
+//
+// See https://github.com/badges/shields/issues/3236 for more information.
+//
+// This service should gradually be replaced by services that handle
+// their respective metrics by providing badges with more context
+// (i.e. a red/error badge when there are multiple security issues)
+//
+// https://docs.sonarqube.org/latest/user-guide/metric-definitions
+const complexityMetricNames = ['complexity', 'cognitive_complexity']
+const duplicationMetricNames = [
+  'duplicated_blocks',
+  'duplicated_files',
+  'duplicated_lines',
+  'duplicated_lines_density',
+]
+// Sonar seemingly has used the terms 'issues' and 'violations' interchangeably
+// so it's possible users are using both/either for badges
+const issuesMetricNames = [
+  'new_violations',
+  'new_blocker_violations',
+  'new_critical_violations',
+  'new_major_violations',
+  'new_minor_violations',
+  'new_info_violations',
+  'blocker_issues',
+  'critical_issues',
+  'major_issues',
+  'minor_issues',
+  'info_issues',
+  'false_positive_issues',
+  'open_issues',
+  'confirmed_issues',
+  'reopened_issues',
+]
+const maintainabilityMetricNames = [
+  'code_smells',
+  'new_code_smells',
+  'sqale_rating',
+  'sqale_index',
+  'sqale_index',
+  'new_technical_debt',
+  'new_sqale_debt_ratio',
+]
+const reliabilityMetricNames = [
+  'bugs',
+  'new_bugs',
+  'reliability_rating',
+  'reliability_remediation_effort',
+  'new_reliability_remediation_effort',
+]
+const securityMetricNames = [
+  'vulnerabilities',
+  'new_vulnerabilities',
+  'security_rating',
+  'security_remediation_effort',
+  'new_security_remediation_effort',
+]
+const sizeMetricNames = [
+  'classes',
+  'comment_lines',
+  'comment_lines_density',
+  'directories',
+  'files',
+  'lines',
+  'nloc',
+  'projects',
+  'statements',
+]
+const testsMetricNames = [
+  'branch_coverage',
+  'new_branch_coverage',
+  'branch_coverage_hits_data',
+  'conditions_by_line',
+  'covered_conditions_by_line',
+  'new_coverage',
+  'line_coverage',
+  'new_line_coverage',
+  'coverage_line_hits_data',
+  'lines_to_cover',
+  'new_lines_to_cover',
+  'skipped_tests',
+  'uncovered_conditions',
+  'new_uncovered_conditions',
+  'uncovered_lines',
+  'new_uncovered_lines',
+  'tests',
+  'test_execution_time',
+  'test_errors',
+  'test_failures',
+  'test_success_density',
```

Comment:
Do we actually need to explicitly whitelist every possible metric value here?

If the _only_ thing we're gaining here is a more specific regexion the URL, I'd be happy to just accept any string in the URL schema and throw a `NotFound()` or something if the key gives us an error

Reply:
This may very well have been an unrelated user error on my part, but I was having some routing issues locally and I thought it may have been due to having a conflict between the services that explicitly declare their metric(s) (like `coverage`) and an identical route here but with a catch all metric. 

i.e.:
`sonar/:protocol(http|https)/:host(.+)/:component(.+)/coverage`

and 

`sonar/:protocol(http|https)/:host(.+)/:component(.+)/:metric`

since we should only hit the generic catch all if the specific metric is not defined explicitly in another service class' route

Reply:
Right yeah I see the problem. We want to avoid making a breaking change and need to avoid defining two conflicting or ambiguous routes :/

Reply:
Exactly. I think my personal preference would be to remove this generic/catch-all support, even if it does potentially break anyone that's using one of those metrics. 

I'm struggling to see/understand the use case where this would be used, since it's basically just a pass through and ends up with badges that don't make a ton of sense:

![](https://img.shields.io/badge/Security%20Rating-F-brightgreen.svg)

If we want to maintain BC support for those badges, then this looks approach seems likely to be the "least bad" solution from my perspective. I suppose the silver lining is that this generic service is just a temporary/tactical solution until we can implement more traditional service classes for the other metrics

Reply:
I think what we've done here is fine, but just reading your ~crossed out~ comments and merging in the conversation in https://github.com/badges/shields/pull/3189#discussion_r275127813 , I agree it would be nice not to have to explicitly take on support for the entire kitchen sink of Sonar functionality. Here's another idea to consider before we merge:

Although we theoretically allow _every sonar metric ever_ , in reality there is probably a relatively small subset of them in use in the real world. If we don't really want to support all of them, we could merge this but not include any examples for `SonarGeneric`. Once we convert the badge to the new service class style, we should be able to start picking up metrics for it (legacy badge traffic doesn't get picked up by prometheus). When we do that, we'll be able to understand which metrics are actually getting any real-word usage and then we could constrain the `SonarGeneric` to only support the metrics which get any traffic before we add it to the index page.

Reply:
I'm good with that too

Reply:
OK. I'll defer to your judgement to either keep the examples in or remove them with the intention that we can look at the (Prometheus) metrics and decide later.

Beyond that, I think everything else is resolved and this is the last thing outstanding, so I'll aim to come back to this and give it one final read over tomorrow evening.

Reply:
Sorry I'm multi tasking but should have been more explicit.

I like your proposal and will make that change shortly. I think your hypothesis is correct as well, and I'd guess there's only a handful of "other" sonar metrics being used... although I'm regularly surprised by what the Shields metrics show 😄